osdep/win32/smtc.cc

```diff
@@ -0,0 +1,389 @@
+        auto &event = *reinterpret_cast<mpv_event*>(wParam);
+
+        update_state(smtc, mpv);
+        if (event.event_id == MPV_EVENT_PROPERTY_CHANGE) {
+            auto &prop = *static_cast<mpv_event_property *>(event.data);
+            if (!strcmp(prop.name, "time-pos"))
+                return 0;
+        }
+        update_metadata(smtc, mpv);
+
+        return 0;
+    }
+
+    if (uMsg == WM_DESTROY)
+        PostQuitMessage(0);
+
+    return DefWindowProc(hWnd, uMsg, wParam, lParam);
+}
+
+static MP_THREAD_VOID win_event_loop_fn(void *arg)
+{
+    mp_thread_set_name("smtc/win");
+    auto &ctx = *static_cast<smtc_ctx*>(arg);
+
+    WNDCLASS wc = {
+        .lpfnWndProc = WindowProc,
+        .hInstance = HINST_THISCOMPONENT,
+        .hIcon = LoadIconW(HINST_THISCOMPONENT, L"IDI_ICON1"),
+        .lpszClassName = L"mpv-smtc"
+    };
+    RegisterClassW(&wc);
+
+    try {
+        ctx.hwnd = CreateWindowExW(0, wc.lpszClassName, L"mpv smtc",
+                                   WS_OVERLAPPEDWINDOW, CW_USEDEFAULT,
+                                   CW_USEDEFAULT, CW_USEDEFAULT, CW_USEDEFAULT,
+                                   nullptr, nullptr, wc.hInstance, nullptr);
+        if (!ctx.hwnd)
+            winrt::throw_last_error();
+
+        SystemMediaTransportControls smtc{ nullptr };
+        auto interop = winrt::get_activation_factory<SystemMediaTransportControls,
+                                                     ISystemMediaTransportControlsInterop>();
+        HRESULT hr = interop->GetForWindow(ctx.hwnd,
+                                           winrt::guid_of<SystemMediaTransportControls>(),
+                                           winrt::put_abi(smtc));
+        if (FAILED(hr))
+            winrt::throw_hresult(hr);
+        SetWindowLongPtrW(ctx.hwnd, GWLP_USERDATA, (LONG_PTR)&smtc);
+
+        smtc.IsEnabled(true);
+
+        smtc.ButtonPressed([&](const SystemMediaTransportControls &,
+                               const SystemMediaTransportControlsButtonPressedEventArgs &args) {
+            const auto pause = [&](int p) {
+                mpv_set_property(ctx.mpv, "pause", MPV_FORMAT_FLAG, &p);
+            };
+            switch (args.Button()) {
+            case SystemMediaTransportControlsButton::Play:
+                pause(false);
+                break;
+            case SystemMediaTransportControlsButton::Pause:
+                pause(true);
+                break;
+            case SystemMediaTransportControlsButton::Stop:
+                mpv_command_string(ctx.mpv, "stop");
+                break;
+            case SystemMediaTransportControlsButton::Next: {
+                int64_t ch_index = -1, ch_count = -1;
+                mpv_get_property(ctx.mpv, "chapter", MPV_FORMAT_INT64, &ch_index);
+                mpv_get_property(ctx.mpv, "chapter-list/count", MPV_FORMAT_INT64, &ch_count);
+                mpv_command_string(ctx.mpv, ch_index < ch_count ? "add chapter 1" : "playlist-next");
```

Comment:
I have a problem with a video that has no chapters: when clicking "Next" in Windows Media Control, it doesn't go to the next file (even if the next file exists in the playlist). Maybe there's a problem here?

Reply:
yep, probably count is 0 in this case.